src/ch/ch_monitor.c

```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+            break;
+        case virCHMonitorVmmEventShutdown: // shutdown inside vmm
+        case virCHMonitorVmEventShutdown:
+            {
+            virDomainState state;
+
+            virObjectLock(vm);
+            state = virDomainObjGetState(vm, NULL);
+            if ((ev == virCHMonitorVmmEventShutdown ||
+                 state == VIR_DOMAIN_SHUTDOWN) &&
+                    (virCHDomainObjBeginJob(vm, CH_JOB_MODIFY) == 0)) {
+                virCHProcessStop(CH_DOMAIN_PRIVATE(vm)->driver,
+                                 vm, VIR_DOMAIN_SHUTOFF_SHUTDOWN);
+                virCHDomainObjEndJob(vm);
+            }
+            virObjectUnlock(vm);
+            break;
+            }
+        case virCHMonitorVmEventBooting:
+        case virCHMonitorVmEventPausing:
+        case virCHMonitorVmEventPaused:
+        case virCHMonitorVmEventResuming:
+        case virCHMonitorVmEventSnapshotting:
+        case virCHMonitorVmEventSnapshotted:
+        case virCHMonitorVmEventRestoring:
+        case virCHMonitorVmEventResizing:
+        case virCHMonitorVmEventResized:
+        case virCHMonitorVmEventDeleted:
+        case virCHMonitorVmmEventStarting:
+        case virCHMonitorCpuCreateVcpu:
+            break;
+        case virCHMonitorEventMax:
+        default:
+            VIR_WARN("unkown event from monitor!");
+            break;
+    }
+
+    return 0;
+}
+
+/*
+ * Helper function to find a block of valid JSON
+ * from a stream of multiple JSON blocks.
+ */
+static inline char *end_of_json(char *str)
+{
+    bool started = false;
+    int blocks = 0;
+    int i = 0;
+
+    while (!started || blocks > 0) {
+        if (str[i] == '{') {
+            if (!started)
+                started = true;
+            blocks++;
+        } else if (str[i] == '}') {
+            blocks--;
+        }
+        i++;
+    }
+
+    return str + i;
+}
+
+/*
+ * Caller should have reference on Monitor and Domain
+ */
+static int virCHMonitorProcessEvents(virCHMonitorPtr mon, int events)
+{
+    virJSONValuePtr obj = NULL;
+    char *buf = mon->buffer;
+    int ret = 0;
+    int i = 0;
+
+    for (i = 0; i < events; i++) {
+        char *end = end_of_json(buf);
+        *end = 0;
+
+        if ((obj = virJSONValueFromString(buf))) {
+            if (virCHMonitorProcessEvent(mon, obj) < 0) {
+                VIR_WARN("Failed to process the event!");
+                ret = -1;
+            }
+            virJSONValueFree(obj);
+        } else {
+            VIR_WARN("Invalid JSON from monitor");
+            ret = -1;
+        }
+
+        *end = '{';
+        buf = end;
+    }
+
+    return ret;
+}
+
+static int virCHMonitorReadProcessEvents(virCHMonitorPtr mon)
+{
+    virDomainObjPtr vm = mon->vm;
+    int events = 0;
+    ssize_t sz = 0;
+    bool incomplete;
+
+    memset(mon->buffer, 0, CH_MONITOR_BUFFER_SZ);
+    do {
+        int ret;
+
+        ret = read(mon->monitor_fd, mon->buffer + sz,
+                   CH_MONITOR_BUFFER_SZ);
```

Comment:
This could read past the end of the `mon->buffer` when `sz > 0`.

Reply:
Thanks for catching this. Will fix!

Reply:
Fixed